docs/content/manual/manual.yml

```diff
@@ -2924,32 +2944,72 @@ sections:
+          The form is `reduce EXP as $var (INIT; UPDATE)`.
+          As an example, we'll pass `[1,2,3]` to this expression:
+
+              reduce .[] as $item (0; . + $item)
+
+          For each result that `.[]` produces, `. + $item` is run to
+          accumulate a running total, starting from 0 as the input value.
+          In this example, `.[]` produces the results 1, 2, and 3, so the
+          effect is similar to running something like this:
+
+              0 | (1 as $item | . + $item) |
+                  (2 as $item | . + $item) |
+                  (3 as $item | . + $item)
+
+        examples:
+          - program: 'reduce .[] as $item (0; . + $item)'
+            input: '[1,2,3,4,5]'
+            output: ['15']
+
+          - program: 'reduce .[] as [$i,$j] (0; . + $i * $j)'
+            input: '[[1,2],[3,4],[5,6]]'
+            output: ['44']
+
+          - program: 'reduce .[] as {$x,$y} (null; .x += $x | .y += [$y])'
+            input: '[{"x":"a","y":1},{"x":"b","y":2},{"x":"c","y":3}]'
+            output: ['{"x":"abc","y":[1,2,3]}']
+
       - title: "`foreach`"
         body: |
 
           The `foreach` syntax is similar to `reduce`, but intended to
           allow the construction of `limit` and reducers that produce
-          intermediate results (see example).
+          intermediate results.
 
           The form is `foreach EXP as $var (INIT; UPDATE; EXTRACT)`.
-          Like `reduce`, `INIT` is evaluated once to produce a state
-          value, then each output of `EXP` is bound to `$var`, `UPDATE`
-          is evaluated for each output of `EXP` with the current state
-          and with `$var` visible.  Each value output by `UPDATE`
-          replaces the previous state.  Finally, `EXTRACT` is evaluated
-          for each new state to extract an output of `foreach`.
+          As an example, we'll pass `[1,2,3]` to this expression:
+
+              foreach .[] as $item (0; . + $item; . * 2)
 
-          This is mostly useful only for constructing `reduce`- and
-          `limit`-like functions.  But it is much more general, as it
-          allows for partial reductions (see the example below).
+          Like the `reduce` syntax, `. + $item` is run for each result
+          that `.[]` produces, but `. * 2` is run for each intermediate
+          values. In this example, since the intermediate values are
+          1, 3, and 6, the `foreach` expression produces 2, 6, and 12.
+
+          When `EXTRACT` is omitted, the identity filter is used.
 
         examples:
-          - program: '[foreach .[] as $item
-                        ([[],[]];
-                        if $item == null then [[],.[0]] else [(.[0] + [$item]),[]] end;
-                        if $item == null then .[1] else empty end)]'
-            input: '[1,2,3,4,null,"a","b",null]'
-            output: ['[[1,2,3,4],["a","b"]]']
+          - program: 'foreach .[] as $item (0; . + $item)'
```

Comment:
Ditto.  We really need to show that `EXTRACT` can `empty` to skip some outputs.

Reply:
Also, I've come to think that I should just have added a `reduce` with the extra `EXTRACT` expression and not called it `foreach`.  My apologies to all for that.

Reply:
@nicowilliams - Don't forget that a fundamental difference between `reduce` and `foreach` (as implemented) is that executing a `break` within a `reduce` generally makes the `reduce` emit empty, e.g.

$ ./jq -n 'label $out | reduce (1, 2, false, 3) as $item (null; if .==false then ., break $out else $item end)'

(I say "generally" because:
```
$ ./jq -M -n 'label $out | reduce (1,2,false, 3) as $item (null; if .==false then .,break $out else . end)'
null
```
)

manual.yml is unclear on all this.  Perhaps you could clarify the behavior that should be documented.

Reply:
@pkoppstein you never execute the `break` in the second case because you never set the reduction state to `$item`.

Reply:
Sorry, I was in a hurry and distracted.  Anyway, the point remains: the documentation does not make it clear whether reduce's actual behavior (empty on `break`) is intended, so unless you think the actual behavior is a bug, I'd like to make it explicit.

(@itchyny - please advise whether for gojq you were just conforming to jq's actual behavior, and whether you think the jq documentation should be clarified.)

Reply:
I don't see any difference of `break` between `reduce` and `foreach`. This is a topic of `break`, and the behavior of using with `reduce` is already documented in the `Breaking out of control structures` section.
```sh
# foreach emits 0, 1, 3 and then breaks on $item == 3
jq -n 'label $out | foreach range(5) as $item (null; if $item > 2 then break $out else debug + $item end)'

# reduce breaks on $item == 3
jq -n 'label $out | reduce range(5) as $item (null; if $item > 3 then break $out else debug + $item end)'
```

Reply:
@itchyny - The section you refer to in manual.yml indicates that `break` can be used to break out of the `reduce`, but does not explicitly say what the result will be.  One might expect the result to be the memoized value at that point, since `reduce` otherwise emits that value.  

(Certainly, one might **want** there to be a way to use `reduce` in such a way, e.g. to short-circuit the computation of the max of a stream, where it is known beforehand that the items in the stream can never be larger than a certain value that may or may not appear in the stream.)